doc: extensions: add custom role 'module_file'

gautesl opened the pull request:
Added another custom role `module_file` which functions like `zephyr_file` but for any module in the west manifest.

The new role must take a module as an argument. Example syntax:

```rst
:module_file:`module-name: path/within/module`
```

or with custom link text:

```rst
:module_file:`my link text <module-name: path/within/module>`
```

Signed-off-by: Gaute Svanes Lunde <gaute.lunde@nordicsemi.no>

gmarull commented on doc/_extensions/zephyr/link-roles.py:
```diff
@@ -21,61 +23,74 @@
+        )
     except subprocess.CalledProcessError:
-        return 'main'
+        return "main"
```
please split formatting changes from actual extension additions

gmarull commented on doc/_extensions/zephyr/link-roles.py:
```diff
@@ -21,61 +23,74 @@
-
-    app.add_role('zephyr_file', autolink('{}/blob/{}/%s'.format(baseurl, rev)))
-    app.add_role('zephyr_raw', autolink('{}/raw/{}/%s'.format(baseurl, rev)))
+    baseurl = "https://github.com/zephyrproject-rtos/zephyr"
```
can't baseurl for Zephyr be determined from the manifest?

gautesl replied:
Not if the manifest does not exist, which is the case for the Zephyr docs. This is only used by the already existing `zephyr_file` and `zephyr_raw` roles

gmarull replied:
Hmm Zephyr has a manifest: https://github.com/zephyrproject-rtos/zephyr/blob/main/west.yml

gautesl replied:
I must have built it wrong locally, I will look into it

gautesl committed:
doc: extension: Format link-roles.py

Formatted `link-roles.py` with Black.

Signed-off-by: Gaute Svanes Lunde <gaute.lunde@nordicsemi.no>

gautesl committed:
doc: extensions: add custom role 'module_file'

Added another custom role `module_file` which functions like
`zephyr_file` but for any module in the west manifest.

The new role must take a module as an argument. Example syntax:

```rst
:module_file:`module-name: path/within/module`
```

or with custom link text:

```rst
:module_file:`my link text <module-name: path/within/module>`
```

This change also moves the hardcoded project name and link to
github to `doc/conf.py` by adding the configuration values
`link_roles_manifest_project` and `link_roles_manifest_baseurl`

Signed-off-by: Gaute Svanes Lunde <gaute.lunde@nordicsemi.no>

gmarull commented on doc/_extensions/zephyr/link-roles.py:
```diff
@@ -68,16 +47,49 @@ def setup(app):
+            baseurl = project.url
+            rev = project.revision
+
+        if baseurl is None or module is None:
```
can this condition ever hold true?

gautesl replied:
`module` can be `None` if using the new role `:module_file:` without a module argument. 
`baseurl` can be `None` given the following circumstances:
- `:module_file:` is used while the `link_roles_default_url` config value is not set (defaults to `None`). 
- The provided module is not found in the manifest (invalid module)

gautesl replied:
Failure tests rewritten to include the case where both a default url and an invalid module are provided

gmarull commented on doc/conf.py:
```diff
@@ -251,6 +251,10 @@
 
+# -- Options for zephyr.link-roles ----------------------------------------
+
+link_roles_default_url = "https://github.com/zephyrproject-rtos/zephyr"
```
I'd name it `link_roles_manifest_baseurl`, this should only be used by the repository containing the manifest.

gautesl replied:
Fixed. Also introduced `link_roles_manifest_project` to get the name of the manifest project.

gautesl commented:
Added another configuration value `link_roles_manifest_project` to properly check for invalid modules. Renamed `link_roles_default_url` to `link_roles_manifest_baseurl`, removed potential spaces from provided modules and links and cleared up the error checking.
